### ISSUE TITLE
High client-side latencies and command timeouts with high-enough RPS

### ISSUE THREAD

**Issue by AndrePeng1** [lettuce-io/lettuce-core#2065]
Hello,
We are observing elevated client-side latencies and plenty of command timeouts when interacting with Redis, and we need help pinpointing why this is the case.

### Application Specs
* Expected Production RPS: 40k
* Redis cluster node type: cache.r6g.4xlarge, 3 shards with 2 replica nodes per shard

Our cluster is hosted in AWS Elasticache.

Package dependency versions:
* org.springframework.data:spring-data-redis: `2.3.0.RELEASE`
* io.lettuce:lettuce-core: `6.1.5.RELEASE`
* org.apache.commons:commons-pool2: `2.4.3`

### Lettuce Setup
We have configured Redis in our Spring Boot application as follows:

```
    @Bean
    public RedisConnectionFactory lettuceConnectionFactory(@Value("${deduplication.redis.host}") String host,
                                                           @Value("${deduplication.redis.commandTimeout.ms}") Long timeoutMilliSec,
                                                           @Value("${threadpool.max.size:500}") int maxLettuceConnections,
                                                           MeterRegistry meterRegistry) {
        RedisClusterConfiguration clusterConfiguration = new RedisClusterConfiguration(Collections.singletonList(host));

        ClusterTopologyRefreshOptions topologyRefreshOptions = ClusterTopologyRefreshOptions.builder()
                .enableAllAdaptiveRefreshTriggers()
                .build();

        ClientOptions clientOptions = ClusterClientOptions.builder()
                .topologyRefreshOptions(topologyRefreshOptions)
                .socketOptions(SocketOptions.builder()
                        .keepAlive(SocketOptions.KeepAliveOptions.builder()
                                .enable(true)
                                .build())
                        .build())
                .build();

        MicrometerOptions micrometerOptions = MicrometerOptions.create();
        ClientResources resources = ClientResources.builder()
                .commandLatencyRecorder(new MicrometerCommandLatencyRecorder(meterRegistry, micrometerOptions)).build();

        GenericObjectPoolConfig poolConfig = new GenericObjectPoolConfig();
        poolConfig.setMaxTotal(maxLettuceConnections);

        LettucePoolingClientConfiguration clientConfig = LettucePoolingClientConfiguration.builder()
                .poolConfig(poolConfig)
                .commandTimeout(Duration.ofMillis(timeoutMilliSec))
                .readFrom(REPLICA_PREFERRED)
                .clientOptions(clientOptions)
                .clientResources(resources)
                .useSsl()
                .build();

        return new LettuceConnectionFactory(clusterConfiguration, clientConfig);
    }

    @Bean(name = "deduplication-cache")
    public RedisTemplate<String, String> redisTemplate(RedisConnectionFactory redisConnectionFactory) {
        RedisTemplate<String, String> template = new RedisTemplate<>();
        template.setConnectionFactory(redisConnectionFactory);
        return template;
    }
```
### Application code

When processing upstream requests, our application calls Redis as follows using the configured `RedisTemplate` above:

```
@Repository
public class DeduplicationRedisRepository implements DeduplicationRepository {

    private final RedisTemplate<String, String> redisClient;

    public DeduplicationRedisRepository(@Qualifier("deduplication-cache") RedisTemplate<String, String> redisClient) {
        this.redisClient = redisClient;
    }

    public Optional<String> getEventId(String key) {
        String metadata = redisClient.opsForValue().get(key);
        return Optional.ofNullable(metadata);
    }

    public void setEventId(String key, String timeStamp, int ttlSeconds) {
        redisClient.opsForValue().set(key, timeStamp, ttlSeconds, TimeUnit.SECONDS);
    }
}
```

Each application request generates a key, then sends a GET request to Redis.  If the key exists, the application finishes request processing; otherwise the application will call two downstream services for tracking purposes, then call SETEX to write that key into our cluster with the given TTL. 

```
    public Mono<Void> track(
            BeaconParams beaconParams,
            String adEventType,
            long timestamp,
            String instance) {
        String deduplicationKey = beaconParams.getId() + ":" + instance + ":" + timestamp;

        if (deduplicationCacheService.isDuplicateEventId(deduplicationKey)) {
            return Mono.empty()
        }

        downstreamService.track(beaconParams.getLineItemId(), beaconParams.getStreamType());
        deduplicationCacheService.storeEventId(deduplicationKey, String.valueOf(timestamp));
        return Mono.empty();
    }
```

### Initial Observations
* Without specifying a connection pool in our `LettuceConnectionFactory` (i.e: using a plain `LettuceClientConfiguration` instead of `LettucePoolingClientConfiguration`), the average p95 GET client-side latency is 571 ms (derived from Lettuce's `firstresponse` metric).  At the same time, the server-side latency from Elasticache for GET-type commands is under 2 ms.  The latency discrepancy is similar with the SETEX commands.  Meanwhile, the application's CPU utilization on its K8s pods is roughly 20%
* When I attach a connection pool to my `LettuceConnectionFactory` (with the exact same code shown above), the client-side latencies do not improve; in fact they increased by 50 - 100 ms.  What is stranger to me is that the application's CPU utilization didn't increase at all; twas still around 20%.  If the connection pool were set up correctly, I would expect the CPU usage to increase and the client-side latencies / command timeouts decreasing.  
* From the longest-lasting Datadog APM Traces against our Redis cluster, these are common errors I see:
-- `javax.net.ssl.SSLException: SSLEngine closed already`
-- `io.netty.handler.ssl.SslHandShakeTimeoutException: handshake timed out after 2000ms`

### Questions
* Is my Lettuce connection pool config setup incorrect?  If not, why does it take our application so long to get a Lettuce connection?
* Do the Spring Redis or Lettuce libraries expose other metrics on the connection pool or how long each executable code path takes?  Would be good for me to understand which code paths inside these libraries my application is spending the most time in

Will comment in this thread with more observations that I find, as I go.

**Comment by mp911de** [comment]
Generally, connection pools are only needed if your application requires the use of more than a single connection. A good reason to have multiple connections is the use of blocking commands (e.g. `BLPOP`) or transactions. Your code above shows neither one.

Also, introducing a connection pool adds overhead and that is why you don't see any improvement, rather additional overhead.

> `javax.net.ssl.SSLException: SSLEngine closed already`

This happens when Redis closes connections upon connect (max connections threshold hit on the server side) or the handshake times out. That issue was kind-of addressed with #2023 by retaining the actual cause on handshake errors.

When using the reactive programming model please keep in mind that all processing of reactive operators happens on the I/O thread. If your operators are CPU-heavy, then the event loop thread will be busy with processing instead of handling I/O traffic. You can enable `publishOnScheduler` via `ClientOptions` to use a scheduler thread instead of the event loop thread for reactive signal dispatch.

There are currently no additional ool metrics exposed by Spring Data Redis.

**Comment by AndrePeng1** [comment]
Thanks @mp911de .  I've found that `publishOnScheduler` does at least reduce how many Redis connection timeout failures the application reports.  Meanwhile, we're also looking to reduce our call volume to Redis, as our current call pattern is somewhat redundant.

This issue can be closed.
